Refactor function spec change detection

stefanprodan opened the pull request:
## Description

- use go-cmp to reliably detect changes in function spec
- log function spec diff before update
- fix deployment annotation append bug by copying the values from function spec
- set the default API namespace the same as the controller default
- change K8s event logger verbosity level to prevent log filling up when a function goes into crash loop, when debugging the event log can be enabled using the -v=3 flag

Fix  #50

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
- [x] I have raised an issue to propose this change ([required](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md))


## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- [ ] I have run the [certifier](https://github.com/openfaas/certifier)

Tested with `stefanprodan/openfaas-operator:fix50v2` on GKE 1.10 by applying changes to every function spec field.

Function definition:

```yaml
apiVersion: openfaas.com/v1alpha2
kind: Function
metadata:
  name: env
spec:
  name: env
  image: stefanprodan/of-env
  annotations:
    test.min: "1"
    test.max: "2"
  labels:
    com.openfaas.scale.min: "2"
    com.openfaas.scale.max: "5"
  environment:
    test.min: "1"
    test.max: "2"
  secrets:
    - faas-token
    - faas-key
  limits:
    cpu: "100m"
    memory: "128Mi"
  requests:
    cpu: "10m"
    memory: "64Mi"
  constraints:
    - "cloud.google.com/gke-nodepool=default-pool"
```

**Tests**

Image change:

```
I0922 15:10:45.008390    4953 deployment.go:229] Change detected for env diff
{v1alpha2.FunctionSpec}.Image:
        -: "stefanprodan/of-env"
        +: "stefanprodan/of-env:1.1.0"
```

Annotation value change:

```
I0922 15:11:44.558130    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Annotations)["test.max"]:
        -: "2"
        +: "3"
```

Annotation key added:

```
I0922 15:13:30.037083    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Annotations)["test.add"]:
        -: <non-existent>
        +: "test"
```

Annotation key removed:

```
I0922 15:14:10.565108    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Annotations)["test.add"]:
        -: "test"
        +: <non-existent>
```

Secret rename:

```
I0922 15:14:59.623583    4953 deployment.go:229] Change detected for env diff
{v1alpha2.FunctionSpec}.Secrets[1]:
        -: "faas-key"
        +: "faas-key2"
```

Label change:

```
I0922 15:16:42.510263    4953 deployment.go:229] Change detected for env diff
(*{v1alpha2.FunctionSpec}.Labels)["com.openfaas.scale.min"]:
        -: "2"
        +: "1"
```

**Note**: annotations, labels, limit, requests and environment have the same type, all tested show the same behaviour. Same with secrets and constraints.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Impact to existing users
<!-- What must existing users do to adopt this change? -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I've read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] I have signed-off my commits with `git commit -s`
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

alexellis commented on pkg/controller/deployment.go:
```diff
@@ -199,121 +209,29 @@ func makeNodeSelector(constraints []string) map[string]string {
+	prevFnSpec := &faasv1.FunctionSpec{}
+	err := json.Unmarshal([]byte(prevFnSpecJson), prevFnSpec)
 	if err != nil {
-		glog.V(2).Infof("Failed to parse previous function spec: %s", err.Error())
```
If all this code is deleted (replaced by go-cmp) then we'll going to do some regression testing on each of the items that can cause a redeployment one by one and in various combinations.

alexellis commented on pkg/server/server.go:
```diff
@@ -23,7 +23,7 @@ const defaultWriteTimeout = 8
 // Start starts HTTP Server for API
 func Start(client clientset.Interface, kube kubernetes.Interface) {
-	functionNamespace := "default"
+	functionNamespace := "openfaas-fn"
```
Seems a valid change, but is this part of the fix?

alexellis commented on pkg/controller/deployment.go:
```diff
@@ -148,18 +149,27 @@ func makeLabels(function *faasv1.Function) map[string]string {
-	annotations["prometheus.io.scrape"] = "false"
+	// save function spec in deployment annotations
+	// used to detect changes in function spec
+	specJSON, err := json.Marshal(function.Spec)
```
Can you confirm that we can no longer get the bug with the infinite annotation? Would be good to see a test showing that the bug is present in the old code and then see it fixed in the new code. 

`makeAnnotations` should be easy to add tests for.

stefanprodan replied:
I've manually tested changes to secrets, annotations and labels (keys and values), image, environment, limits and requests. The JSON annotation was updated not appended and all changes were detected correctly.

alexellis commented on pkg/controller/deployment.go:
```diff
@@ -199,121 +209,29 @@ func makeNodeSelector(constraints []string) map[string]string {
 
 // deploymentNeedsUpdate determines if the function spec is different from the deployment spec
```
Would be ideal to see a test on `deploymentNeedsUpdate`

alexellis commented:
This will need some exploration testing with adding / removing and updating each field covered by the spec.

stefanprodan commented:
@alexellis I've tested all the function spec fields: secrets (add, delete, rename), annotations and labels (keys and values), image, environment, limits, requests and constraints. I've also added tests for make annotations func. I've updated the PR description with some examples.

alexellis commented on pkg/controller/annotations_test.go:
```diff
@@ -0,0 +1,70 @@
+	faasv1 "github.com/openfaas-incubator/openfaas-operator/pkg/apis/openfaas/v1alpha2"
+)
+
+func Test_makeAnnotations_NoKeys(t *testing.T) {
```
This I s basically the tests from my gist? Any other scenarios we should cover? cc @rgee0

johnmccabe replied:
Setting `prometheus.io.scrape` up front in the `annotationVal` as a `string`, `bool`, number, `null`
